# 1. Importing Rust Library

In [1]:
import my_rust_module

# 2. Defining the same function which is implemented in the Rust module

In [7]:
def fib(n: int) -> int:
    a, b = 0, 1
    for _ in range(n):
        a, b = b, a + b
    return a

# 3. Comparing the compilation time of Rust binding vs Python

In [6]:
%timeit fib(90)
%timeit my_rust_module.fib(90)

1.35 μs ± 7.22 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
629 ns ± 7.74 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


## Performance Comparison: Python vs Rust (PyO3)

To compare the execution speed of Python and Rust, the same iterative algorithm was implemented in both languages to compute the **nth Fibonacci number**. Both functions return only a single integer, ensuring that the benchmark primarily measures the computation itself.

The benchmark was performed using Jupyter's `%timeit` magic command.

| Implementation | Execution Time |
|----------------|---------------:|
| Python | **1.35 μs ± 7.22 ns** |
| Rust (PyO3) | **629 ns ± 7.74 ns** |

The Rust implementation is approximately **2.1× faster** than the equivalent Python implementation.

When a Rust function is called from Python using PyO3, the execution involves more than just running the Rust code. Python first transfers control to the compiled Rust extension, where the input arguments are converted from Python objects to their corresponding Rust types. After the computation, the Rust return value is converted back into a Python object before control is returned to the Python interpreter. This process introduces a fixed overhead for every function call, commonly referred to as the **Foreign Function Interface (FFI) overhead**.

For this benchmark, the overhead is relatively small because only a single integer is passed to Rust and a single integer is returned. If the function instead returned an entire list of Fibonacci numbers, every Rust integer would need to be converted into a Python integer and inserted into a Python list, substantially increasing the data conversion cost. Consequently, returning a single value provides a more accurate comparison of the computational performance of the Python and Rust implementations, while minimizing the influence of data transfer between the two languages.